ls1intum/Artemis
# [Feature] Exam Archival

### Checklist

### Motivation and Context


It should be possible for instructors to archive/export an exam after it is over. Archiving an exam compresses all student submissions for each exam exercise into a single zip file (similar to the existing course archival functionality).

### Description

This PR implements the archival of a specific exam. It does this by:

1. Extending the existing export functionalities in `CourseExportService` to support exporting specific exams. The service is renamed to `CourseExamExportService` since they share the same implementation.
2. Adding to new endpoints in `ExamResource` that enable instructors to archive an exam and download it.
3. Adds a new Angular component that implements the archive and download buttons for courses and exams
4. Implements the archive functionality in the exam details component

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create an exam with Text, File Upload, Modeling, and Programming exercises
4. Participate in the exam with a couple of users by generating submissions
5. During the exam and as an instructor, navigate to Course Administration -> Exams -> the exam
6. Make sure the archive button isn't visible
7. When the exam finishes, you should be able to archive the exam. Click on the button and wait until it's done
8. A download button should appear and you should be able to download the zip file.
9. Make sure the zip files contains all student submissions for each exercise
10. You can also log in as a teaching assistant and make sure that you don't have access to the archive button (only instructors and admins have access to it)

### Test Coverage



CourseExamExportService.java 80%
CourseService.java 93%
GroupNotificationService.java 70%
ExamService.java 93%
ExamResource.java 90%
course-detail.component.ts 67%
exam-management.service.ts 66%
exam-detail.component.ts 100%
course-exam-archive-button.component.ts 90%

### Screenshots


The archive button appears when the exam is over:
![image](https://user-images.githubusercontent.com/11298674/111661345-3f21d180-880f-11eb-97df-496e17a96a49.png)

A warning is displayed when pressing the archive button:
![image](https://user-images.githubusercontent.com/11298674/111661477-55c82880-880f-11eb-8681-5bc62722ac75.png)

It's possible to download the zip file once archving is done:
![image](https://user-images.githubusercontent.com/11298674/111663404-0f73c900-8811-11eb-8f99-51b2a593922d.png)
